File: data_cleaner.py
"""
Data Cleaning Module for Bank Transaction Data

This module handles normalization and cleaning of extracted transaction data,
including currency formatting, column normalization, and handling multi-line descriptions.
"""

import pandas as pd
import re
from typing import Optional, Any


def normalize_columns(df: pd.DataFrame) -> pd.DataFrame:
    """
    Standardize column names by stripping whitespace and converting to uppercase.
    
    Args:
        df: DataFrame with potentially inconsistent column names
        
    Returns:
        DataFrame with normalized column names
    """
    df = df.copy()
    
    # Normalize column names
    df.columns = [str(col).strip().upper() if col else f"COL_{i}" 
                  for i, col in enumerate(df.columns)]
    
    return df


def clean_currency_value(value: Any) -> Optional[float]:
    """
    Clean currency values by removing currency symbols and thousand separators.
    Handles Indonesian format: 1.000.000,00 or 98,779,762.35
    Also handles DB/CR indicators that may be attached to amounts.
    
    Args:
        value: Currency value as string or number
        
    Returns:
        float: Numeric value, or None if conversion fails
    """
    if pd.isna(value) or value is None:
        return None
    
    # Convert to string
    value_str = str(value).strip()
    
    if not value_str or value_str == "":
        return None
    
    # Remove currency symbols (Rp, $, etc.)
    value_str = re.sub(r'[Rr][Pp]\s*', '', value_str)
    value_str = re.sub(r'[$€£¥]', '', value_str)
    
    # Remove DB/CR indicators (Debit/Credit indicators)
    value_str = re.sub(r'\s*(DB|CR|DEBIT|CREDIT)\s*$', '', value_str, flags=re.IGNORECASE)
    value_str = re.sub(r'\s*(DB|CR|DEBIT|CREDIT)\s*', ' ', value_str, flags=re.IGNORECASE)
    
    # Handle BCA format: numbers like "98,779,762.35" or "23,400.00"
    # Check if it uses comma as thousand separator and dot as decimal
    if ',' in value_str and '.' in value_str:
        # Format like "98,779,762.35" - comma is thousand separator, dot is decimal
        # Remove commas, keep dot
        value_str = value_str.replace(',', '')
    elif ',' in value_str:
        # Only comma present - could be decimal separator (Indonesian format) or thousand
        # Check if there are multiple commas (thousand separators)
        comma_count = value_str.count(',')
        if comma_count > 1 or (comma_count == 1 and len(value_str.split(',')[1]) <= 2):
            # Likely decimal separator (e.g., "1.000.000,50")
            # First remove dots (thousand separators), then replace comma with dot
            value_str = value_str.replace('.', '')
            value_str = value_str.replace(',', '.')
        else:
            # Single comma, might be thousand separator or decimal
            # If digits after comma <= 2, it's likely decimal separator
            parts = value_str.split(',')
            if len(parts) == 2 and len(parts[1]) <= 2:
                value_str = parts[0].replace('.', '') + '.' + parts[1]
            else:
                # Treat as thousand separator
                value_str = value_str.replace(',', '')
    else:
        # Only dots - could be thousand separators (Indonesian format)
        # If there are multiple dots, they're thousand separators
        dot_count = value_str.count('.')
        if dot_count > 1:
            # Multiple dots = thousand separators, remove them
            value_str = value_str.replace('.', '')
        # If single dot, keep it as decimal separator
    
    # Remove any remaining non-numeric characters except decimal point and minus sign
    value_str = re.sub(r'[^\d.\-]', '', value_str)
    
    try:
        return float(value_str) if value_str else None
    except (ValueError, TypeError):
        return None


def merge_multiline_descriptions(df: pd.DataFrame) -> pd.DataFrame:
    """
    Handle KETERANGAN and DETAIL TRANSAKSI fields that may span multiple lines.
    Rows without a date in TANGGAL are assumed to be continuations of the previous row.
    Updated for BCA format with DD/MM date pattern.
    
    Args:
        df: DataFrame with transaction data
        
    Returns:
        DataFrame with merged multi-line descriptions
    """
    if df.empty:
        return df
    
    df = df.copy()
    
    # Identify rows that are likely continuations (no date in first column)
    merged_rows = []
    current_row = None
    
    for idx, row in df.iterrows():
        tanggal = str(row.iloc[0]).strip() if pd.notna(row.iloc[0]) else ""
        
        # Check if this row has a date (DD/MM format for BCA statements)
        # Pattern: 01/08, 1/8, 31/12, etc.
        has_date = bool(re.search(r'\d{1,2}/\d{1,2}', tanggal)) if tanggal else False
        
        if has_date:
            # This is a new transaction row
            if current_row is not None:
                merged_rows.append(current_row)
            current_row = row.copy()
        else:
            # This is a continuation row - merge KETERANGAN and DETAIL TRANSAKSI with previous row
            if current_row is not None:
                # Merge KETERANGAN column (index 1)
                keterangan_col_idx = 1
                if len(current_row) > keterangan_col_idx and len(row) > keterangan_col_idx:
                    prev_keterangan = str(current_row.iloc[keterangan_col_idx]) if pd.notna(current_row.iloc[keterangan_col_idx]) else ""
                    curr_keterangan = str(row.iloc[keterangan_col_idx]) if pd.notna(row.iloc[keterangan_col_idx]) else ""
                    
                    if curr_keterangan.strip():
                        if prev_keterangan:
                            merged_keterangan = f"{prev_keterangan} {curr_keterangan}".strip()
                        else:
                            merged_keterangan = curr_keterangan.strip()
                        current_row.iloc[keterangan_col_idx] = merged_keterangan
                
                # Merge DETAIL TRANSAKSI column (index 2) - continuation details
                detail_col_idx = 2
                if len(current_row) > detail_col_idx and len(row) > detail_col_idx:
                    prev_detail = str(current_row.iloc[detail_col_idx]) if pd.notna(current_row.iloc[detail_col_idx]) else ""
                    curr_detail = str(row.iloc[detail_col_idx]) if pd.notna(row.iloc[detail_col_idx]) else ""
                    
                    if curr_detail.strip():
                        if prev_detail:
                            merged_detail = f"{prev_detail} {curr_detail}".strip()
                        else:
                            merged_detail = curr_detail.strip()
                        current_row.iloc[detail_col_idx] = merged_detail
                
                # If MUTASI or SALDO appear in continuation and are missing from main row, use them
                mutasi_col_idx = 3
                saldo_col_idx = 4
                if len(current_row) > mutasi_col_idx and len(row) > mutasi_col_idx:
                    if (not current_row.iloc[mutasi_col_idx] or str(current_row.iloc[mutasi_col_idx]).strip() == "") and \
                       row.iloc[mutasi_col_idx] and str(row.iloc[mutasi_col_idx]).strip():
                        current_row.iloc[mutasi_col_idx] = row.iloc[mutasi_col_idx]
                
                if len(current_row) > saldo_col_idx and len(row) > saldo_col_idx:
                    if (not current_row.iloc[saldo_col_idx] or str(current_row.iloc[saldo_col_idx]).strip() == "") and \
                       row.iloc[saldo_col_idx] and str(row.iloc[saldo_col_idx]).strip():
                        current_row.iloc[saldo_col_idx] = row.iloc[saldo_col_idx]
    
    # Add the last row
    if current_row is not None:
        merged_rows.append(current_row)
    
    if merged_rows:
        return pd.DataFrame(merged_rows).reset_index(drop=True)
    else:
        return df


def clean_transaction_data(df: pd.DataFrame, expected_columns: list = None) -> pd.DataFrame:
    """
    Apply all cleaning transformations to transaction data.
    
    Args:
        df: Raw DataFrame from PDF extraction
        expected_columns: List of expected column names in order
        
    Returns:
        Cleaned DataFrame ready for display/export
    """
    if df.empty:
        return df
    
    df = df.copy()
    
    # Normalize column names
    df = normalize_columns(df)
    
    # Set expected column names if provided
    if expected_columns and len(df.columns) >= len(expected_columns):
        df.columns = expected_columns[:len(df.columns)] + list(df.columns[len(expected_columns):])
    
    # Remove completely empty rows
    df = df.dropna(how='all')
    
    # Handle multi-line descriptions
    df = merge_multiline_descriptions(df)
    
    # Clean string columns (strip whitespace)
    for col in df.columns:
        if df[col].dtype == 'object':
            df[col] = df[col].apply(lambda x: str(x).strip() if pd.notna(x) else "")
    
    # Clean numeric columns (MUTASI and SALDO)
    numeric_columns = ['MUTASI', 'SALDO']
    for col in numeric_columns:
        if col in df.columns:
            df[col] = df[col].apply(clean_currency_value)
    
    # Remove rows where TANGGAL is empty (invalid transactions)
    if 'TANGGAL' in df.columns:
        df = df[df['TANGGAL'].notna() & (df['TANGGAL'].astype(str).str.strip() != "")]
    
    # Reset index
    df = df.reset_index(drop=True)
    
    return df

File: lib/dataCleaner.ts
import { TransactionRow } from './types';

const EXPECTED_COLUMNS = ["TANGGAL", "KETERANGAN", "DETAIL TRANSAKSI", "MUTASI", "SALDO"];

/**
 * Check if a number string looks like a currency value.
 * Currency values typically have:
 * - Thousand separators (commas or dots)
 * - Decimal places (.XX)
 * - DB/CR indicators
 * - Are reasonably large (> 100 or have decimal places)
 */
export function isCurrencyValue(value: string): boolean {
  if (!value || value.trim().length === 0) {
    return false;
  }

  const trimmed = value.trim();
  
  // Check for DB/CR indicators - these are strong signals of currency
  if (/\b(DB|CR|DEBIT|CREDIT)\b/i.test(trimmed)) {
    return true;
  }

  // Check for thousand separators (commas or dots in number pattern)
  // Pattern: digits, then comma/dot, then 3 digits (repeated), optionally followed by decimal
  if (/\d{1,3}(?:[,\.]\d{3})+(?:[,\.]\d{2})?/.test(trimmed)) {
    return true;
  }

  // Check for decimal places (at least 2 digits after decimal)
  if (/\.\d{2,}/.test(trimmed)) {
    // Extract numeric value to check if it's substantial
    const numericPart = trimmed.replace(/[^\d.]/g, '');
    const numValue = parseFloat(numericPart);
    if (!isNaN(numValue) && numValue >= 1) {
      return true;
    }
  }

  // Check if it's a large number without decimals (>= 1000)
  const numericPart = trimmed.replace(/[^\d]/g, '');
  const numValue = parseFloat(numericPart);
  if (!isNaN(numValue) && numValue >= 1000) {
    return true;
  }

  return false;
}

/**
 * Extract currency values (MUTASI and SALDO) from a text string.
 * Uses right-to-left parsing to find the last two currency values.
 */
export function extractCurrencyValues(line: string): { mutasi: string; saldo: string; remainingLine: string } {
  let mutasi = '';
  let saldo = '';
  let remainingLine = line;

  if (!line || line.trim().length === 0) {
    return { mutasi, saldo, remainingLine };
  }

  // Pattern to match currency values:
  // - Numbers with thousand separators: 23,000.00 or 111,686,562.71
  // - Numbers with DB/CR: 23,000.00 DB or 23,000.00DB
  // - Numbers with decimals: 23.50 or 111686562.71
  // - Large numbers: 1000 or more
  const currencyPattern = /(\d{1,3}(?:[,\.]\d{3})*(?:[,\.]\d{2})?(?:\s*(?:DB|CR))?|\d+\.\d{2,}(?:\s*(?:DB|CR))?|\d{4,}(?:\s*(?:DB|CR))?)/g;
  
  const matches: Array<{ value: string; index: number; length: number }> = [];
  let match;
  
  // Find all potential currency matches
  while ((match = currencyPattern.exec(line)) !== null) {
    const value = match[1].trim();
    // Filter to only include values that look like currency
    if (isCurrencyValue(value)) {
      matches.push({ value, index: match.index, length: match[0].length });
    }
  }

  // Helper: value has DB/CR suffix (transaction amount indicator)
  const hasDbCr = (v: string) => /(DB|CR)/i.test(v);
  // Helper: value looks like running balance (large, no DB/CR - typically 7+ digits)
  const looksLikeBalance = (v: string) => {
    const num = parseFloat(String(v).replace(/[^\d.]/g, ''));
    return !isNaN(num) && num >= 1_000_000 && !hasDbCr(v);
  };
  const toNum = (v: string) => parseFloat(String(v).replace(/[^\d.]/g, '')) || 0;
  const sameAmount = (a: string, b: string) => Math.abs(toNum(a) - toNum(b)) < 0.01;

  // Extract the last two currency values (MUTASI and SALDO)
  if (matches.length >= 2) {
    const lastMatch = matches[matches.length - 1];
    const secondLastMatch = matches[matches.length - 2];
    // Dynamic assignment: DB/CR = MUTASI (transaction amount), large + no DB/CR = SALDO (balance)
    if (hasDbCr(lastMatch.value) && !hasDbCr(secondLastMatch.value)) {
      mutasi = lastMatch.value;
      // Same amount in description (e.g. "3350000.00 kost" + "3,350,000.00DB") → only MUTASI
      saldo = sameAmount(lastMatch.value, secondLastMatch.value) ? '' : (looksLikeBalance(secondLastMatch.value) ? secondLastMatch.value : '');
    } else if (hasDbCr(secondLastMatch.value) && !hasDbCr(lastMatch.value)) {
      mutasi = secondLastMatch.value;
      saldo = lastMatch.value;
    } else if (looksLikeBalance(lastMatch.value) && !looksLikeBalance(secondLastMatch.value)) {
      saldo = lastMatch.value;
      mutasi = secondLastMatch.value;
    } else {
      // Default: second-to-last = MUTASI, last = SALDO
      mutasi = secondLastMatch.value;
      saldo = lastMatch.value;
    }

    // Remove both from the line using their exact indices
    // Sort by index (ascending) to process left to right
    const sortedMatches = [secondLastMatch, lastMatch].sort((a, b) => a.index - b.index);
    
    const parts: string[] = [];
    let lastIndex = 0;
    
    for (const m of sortedMatches) {
      if (m.index > lastIndex) {
        parts.push(line.substring(lastIndex, m.index));
      }
      lastIndex = Math.max(lastIndex, m.index + m.length);
    }
    
    // Add remaining part after the last match
    if (lastIndex < line.length) {
      parts.push(line.substring(lastIndex));
    }
    
    remainingLine = parts.join('').replace(/\s{2,}/g, ' ').trim();
  } else if (matches.length === 1) {
    // Only one currency value - use heuristics: DB/CR = MUTASI, large balance = SALDO
    const singleMatch = matches[0];
    if (hasDbCr(singleMatch.value)) {
      mutasi = singleMatch.value;
    } else if (looksLikeBalance(singleMatch.value)) {
      saldo = singleMatch.value;
    } else {
      // Ambiguous: default to SALDO for backward compat (e.g. SALDO AWAL)
      saldo = singleMatch.value;
    }
    remainingLine = (
      line.substring(0, singleMatch.index) + 
      line.substring(singleMatch.index + singleMatch.length)
    ).replace(/\s{2,}/g, ' ').trim();
  }

  return { mutasi, saldo, remainingLine };
}

/**
 * Clean currency values by removing currency symbols and thousand separators.
 * Handles Indonesian format: 1.000.000,50 or 98,779,762.35
 * Also handles DB/CR indicators that may be attached to amounts.
 */
export function cleanCurrencyValue(value: any): number | null {
  if (value === null || value === undefined || value === '') {
    return null;
  }

  // Convert to string
  let valueStr = String(value).trim();

  if (!valueStr) {
    return null;
  }

  // Remove currency symbols (Rp, $, etc.)
  valueStr = valueStr.replace(/[Rr][Pp]\s*/g, '');
  valueStr = valueStr.replace(/[$€£¥]/g, '');

  // Remove DB/CR indicators (Debit/Credit indicators)
  valueStr = valueStr.replace(/\s*(DB|CR|DEBIT|CREDIT)\s*$/gi, '');
  valueStr = valueStr.replace(/\s*(DB|CR|DEBIT|CREDIT)\s*/gi, ' ');

  // Handle BCA format: numbers like "98,779,762.35" or "23,400.00"
  // Check if it uses comma as thousand separator and dot as decimal
  if (valueStr.includes(',') && valueStr.includes('.')) {
    // Format like "98,779,762.35" - comma is thousand separator, dot is decimal
    // Remove commas, keep dot
    valueStr = valueStr.replace(/,/g, '');
  } else if (valueStr.includes(',')) {
    // Only comma present - could be decimal separator (Indonesian format) or thousand
    // Check if there are multiple commas (thousand separators)
    const commaCount = (valueStr.match(/,/g) || []).length;
    const parts = valueStr.split(',');
    if (commaCount > 1 || (commaCount === 1 && parts[1] && parts[1].length <= 2)) {
      // Likely decimal separator (e.g., "1.000.000,50")
      // First remove dots (thousand separators), then replace comma with dot
      valueStr = valueStr.replace(/\./g, '');
      valueStr = valueStr.replace(',', '.');
    } else {
      // Single comma, might be thousand separator or decimal
      // If digits after comma <= 2, it's likely decimal separator
      if (parts.length === 2 && parts[1].length <= 2) {
        valueStr = parts[0].replace(/\./g, '') + '.' + parts[1];
      } else {
        // Treat as thousand separator
        valueStr = valueStr.replace(/,/g, '');
      }
    }
  } else {
    // Only dots - could be thousand separators (Indonesian format)
    // If there are multiple dots, they're thousand separators
    const dotCount = (valueStr.match(/\./g) || []).length;
    if (dotCount > 1) {
      // Multiple dots = thousand separators, remove them
      valueStr = valueStr.replace(/\./g, '');
    }
    // If single dot, keep it as decimal separator
  }

  // Remove any remaining non-numeric characters except decimal point and minus sign
  valueStr = valueStr.replace(/[^\d.\-]/g, '');

  try {
    return valueStr ? parseFloat(valueStr) : null;
  } catch {
    return null;
  }
}

/**
 * Normalize column names by stripping whitespace and converting to uppercase.
 */
export function normalizeColumns(rows: any[][]): string[] {
  if (!rows || rows.length === 0) {
    return EXPECTED_COLUMNS;
  }

  // Use expected columns if available
  return EXPECTED_COLUMNS;
}

/**
 * Check if a row contains transaction data.
 */
export function isValidTransactionRow(row: any[]): boolean {
  if (!row || row.length < 2) {
    return false;
  }

  // Skip header rows (first cell contains "TANGGAL")
  const firstCell = row[0] ? String(row[0]).trim().toUpperCase() : "";
  if (firstCell.includes("TANGGAL")) {
    return false;
  }

  // Skip completely empty rows
  if (row.every(cell => cell === null || cell === undefined || String(cell).trim() === "")) {
    return false;
  }

  // Check if first column contains a date pattern (DD/MM format)
  const firstCellStr = row[0] ? String(row[0]).trim() : "";
  const datePattern = /\d{1,2}\/\d{1,2}/.test(firstCellStr);

  // Accept rows that either:
  // 1. Have a date in first column (main transaction row)
  // 2. Have content in KETERANGAN or DETAIL TRANSAKSI column (continuation row)
  if (datePattern) {
    return true;
  }

  // Check if it's a continuation row with content
  const secondCellStr = row[1] ? String(row[1]).trim() : "";
  const thirdCellStr = row[2] ? String(row[2]).trim() : "";
  const mutasiCell = row[3] ? String(row[3]).trim() : "";
  const saldoCell = row[4] ? String(row[4]).trim() : "";

  // Accept amounts-only continuation lines (MUTASI/SALDO on separate line in BCA PDFs)
  if (mutasiCell || saldoCell) {
    return true;
  }
  return !!(secondCellStr || thirdCellStr);
}

/**
 * Merge multi-line transactions into single rows.
 */
export function mergeTransactionLines(rows: any[][]): any[][] {
  if (!rows || rows.length === 0) {
    return [];
  }

  const mergedRows: any[][] = [];
  let currentTransaction: any[] | null = null;
  let bufferedAmountsRow: any[] | null = null;

  for (const row of rows) {
    // Defensive padding: ensure row has 5 columns before accessing indices
    const padded = [...row];
    while (padded.length < 5) {
      padded.push('');
    }

    const tanggal = String(padded[0] || '').trim();

    // Check if this row starts with a date (DD/MM format)
    const hasDate = /\d{1,2}\/\d{1,2}/.test(tanggal);

    if (hasDate) {
      // This is a new transaction row
      if (currentTransaction !== null) {
        mergedRows.push(currentTransaction);
      }

      // Start new transaction
      currentTransaction = [...padded];

      // Apply buffered amounts-only row (e.g. after page break) if this row is missing them
      if (bufferedAmountsRow && (!currentTransaction[3] || !String(currentTransaction[3]).trim() || !currentTransaction[4] || !String(currentTransaction[4]).trim())) {
        if (bufferedAmountsRow[3] && String(bufferedAmountsRow[3]).trim()) {
          currentTransaction[3] = bufferedAmountsRow[3];
        }
        if (bufferedAmountsRow[4] && String(bufferedAmountsRow[4]).trim()) {
          currentTransaction[4] = bufferedAmountsRow[4];
        }
        bufferedAmountsRow = null;
      }
    } else {
      // This is a continuation row - merge into current transaction
      if (currentTransaction !== null) {
        // Merge DETAIL TRANSAKSI column (index 2) - continuation details
        const detailValue = padded[2] ? String(padded[2]).trim() : "";
        if (detailValue) {
          const currentDetail = currentTransaction[2] ? String(currentTransaction[2]).trim() : "";
          currentTransaction[2] = currentDetail ? `${currentDetail} ${detailValue}` : detailValue;
        }

        // Also merge KETERANGAN if it has content (index 1)
        const keteranganValue = padded[1] ? String(padded[1]).trim() : "";
        if (keteranganValue) {
          const currentKeterangan = currentTransaction[1] ? String(currentTransaction[1]).trim() : "";
          currentTransaction[1] = currentKeterangan ? `${currentKeterangan} ${keteranganValue}` : keteranganValue;
        }

        // If MUTASI or SALDO appear in continuation, they might be missing from main row
        const mutasiValue = padded[3] ? String(padded[3]).trim() : "";
        if (mutasiValue && (!currentTransaction[3] || !String(currentTransaction[3]).trim())) {
          currentTransaction[3] = mutasiValue;
        }

        const saldoValue = padded[4] ? String(padded[4]).trim() : "";
        if (saldoValue && (!currentTransaction[4] || !String(currentTransaction[4]).trim())) {
          currentTransaction[4] = saldoValue;
        }
      } else {
        // Orphaned amounts-only row (e.g. after page break) - buffer for next transaction
        const hasAmounts = (padded[3] && String(padded[3]).trim()) || (padded[4] && String(padded[4]).trim());
        const isAmountsOnly = hasAmounts && !padded[1] && !padded[2];
        if (isAmountsOnly) {
          bufferedAmountsRow = padded;
        }
      }
    }
  }

  // Add the last transaction
  if (currentTransaction !== null) {
    mergedRows.push(currentTransaction);
  }

  return mergedRows;
}

/**
 * Clean transaction data and convert to structured format.
 */
export function cleanTransactionData(rawRows: any[][]): TransactionRow[] {
  if (!rawRows || rawRows.length === 0) {
    return [];
  }

  // Filter out invalid rows
  const validRows = rawRows.filter(isValidTransactionRow);

  // Merge multi-line transactions
  const mergedRows = mergeTransactionLines(validRows);

  // Convert to TransactionRow format
  const cleanedData: TransactionRow[] = mergedRows
    .map((row) => {
      // Ensure row has at least 5 columns
      const paddedRow = [...row];
      while (paddedRow.length < 5) {
        paddedRow.push(null);
      }
      return {
        TANGGAL: paddedRow[0] ? String(paddedRow[0]).trim() : "",
        KETERANGAN: paddedRow[1] ? String(paddedRow[1]).trim() : "",
        "DETAIL TRANSAKSI": paddedRow[2] ? String(paddedRow[2]).trim() : "",
        MUTASI: cleanCurrencyValue(paddedRow[3]),
        SALDO: cleanCurrencyValue(paddedRow[4]),
      };
    })
    .filter((row) => {
      // Remove rows where TANGGAL is empty (invalid transactions)
      return row.TANGGAL && row.TANGGAL.trim() !== "";
    });

  // Post-processing: extract MUTASI/SALDO from DETAIL TRANSAKSI if still missing.
  // This catches cases where currency values ended up in the description after merging
  // continuation lines (common with pdf-parse text extraction).
  for (const row of cleanedData) {
    if (row.MUTASI === null || row.SALDO === null) {
      const detail = row["DETAIL TRANSAKSI"];
      if (detail) {
        const { mutasi, saldo, remainingLine } = extractCurrencyValues(detail);
        
        if (mutasi && row.MUTASI === null) {
          row.MUTASI = cleanCurrencyValue(mutasi);
        }
        if (saldo && row.SALDO === null) {
          row.SALDO = cleanCurrencyValue(saldo);
        }
        // Update the detail text to remove extracted currency values
        if (mutasi || saldo) {
          row["DETAIL TRANSAKSI"] = remainingLine;
        }
      }
    }
  }

  return cleanedData;
}
